File: apps/android/app/src/main/java/io/github/andivalsei/andivum/SupabaseAuthClient.kt
package io.github.andivalsei.andivum

import java.net.HttpURLConnection
import java.net.URL
import org.json.JSONObject

data class AuthOperationResult(
    val sessionCreated: Boolean,
    val emailConfirmationRequired: Boolean,
)

class SupabaseAuthClient(
    private val supabaseUrl: String,
    private val publishableKey: String,
    private val tokenStore: AuthTokenStore,
    private val openConnection: (URL) -> HttpURLConnection = { url ->
        url.openConnection() as HttpURLConnection
    },
    private val clock: () -> Long = { System.currentTimeMillis() / 1000 },
) {
    private val authBaseUrl = "${supabaseUrl.trimEnd('/')}/auth/v1"
    private val profileClient = SupabaseProfileClient(
        supabaseUrl,
        publishableKey,
        openConnection,
    )

    fun currentSession(): SupabaseTokenSet? = tokenStore.read()

    fun signIn(email: String, password: String): AuthOperationResult =
        authenticate("token?grant_type=password", email, password)

    fun signUp(email: String, password: String): AuthOperationResult =
        authenticate("signup", email, password)

    fun getCurrentSession(): SessionResponse {
        var token = tokenStore.read()
            ?: throw IllegalStateException("No saved session is available.")
        var refreshed = false

        if (needsRefresh(token)) {
            token = refresh(token)
            refreshed = true
        }

        repeat(2) {
            try {
                return profileClient.getCurrentSession(token.accessToken)
            } catch (exception: SessionUnauthorizedException) {
                if (refreshed || token.refreshToken.isNullOrBlank()) {
                    throw exception
                }
                token = refresh(token)
                refreshed = true
            }
        }

        throw IllegalStateException("Supabase session validation could not be completed.")
    }

    fun clearSession() {
        tokenStore.clear()
    }

    fun signOut() {
        val token = tokenStore.read()
        try {
            if (token != null && token.accessToken.isNotBlank()) {
                request(
                    endpoint = "$authBaseUrl/logout",
                    method = "POST",
                    accessToken = token.accessToken,
                )
            }
        } finally {
            tokenStore.clear()
        }
    }

    private fun authenticate(
        path: String,
        email: String,
        password: String,
    ): AuthOperationResult {
        require(email.isNotBlank()) { "Email is required." }
        require(password.isNotBlank()) { "Password is required." }

        val body = JSONObject()
            .put("email", email)
            .put("password", password)
        val response = request(
            endpoint = "$authBaseUrl/$path",
            method = "POST",
            body = body.toString(),
        )
        ensureSuccess(response)
        val json = JSONObject(response.body)
        val accessToken = json.optString("access_token")
        if (accessToken.isBlank()) {
            return AuthOperationResult(
                sessionCreated = false,
                emailConfirmationRequired = path == "signup",
            )
        }

        tokenStore.write(
            SupabaseTokenSet(
                accessToken = accessToken,
                refreshToken = json.optString("refresh_token").takeIf { it.isNotBlank() },
                tokenType = json.optString("token_type", "Bearer"),
                expiresIn = json.optInt("expires_in", 3600),
                issuedAtEpochSeconds = clock(),
            ),
        )
        return AuthOperationResult(
            sessionCreated = true,
            emailConfirmationRequired = false,
        )
    }

    private fun refresh(current: SupabaseTokenSet): SupabaseTokenSet {
        val refreshToken = current.refreshToken
            ?: throw IllegalStateException("No Supabase refresh token is available.")
        val response = request(
            endpoint = "$authBaseUrl/token?grant_type=refresh_token",
            method = "POST",
            body = JSONObject().put("refresh_token", refreshToken).toString(),
        )
        ensureSuccess(response)
        val json = JSONObject(response.body)
        val accessToken = json.optString("access_token")
        require(accessToken.isNotBlank()) { "Supabase refresh returned no access token." }
        val refreshed = SupabaseTokenSet(
            accessToken = accessToken,
            refreshToken = json.optString("refresh_token")
                .takeIf { it.isNotBlank() }
                ?: current.refreshToken,
            tokenType = json.optString("token_type", "Bearer"),
            expiresIn = json.optInt("expires_in", 3600),
            issuedAtEpochSeconds = clock(),
        )
        tokenStore.write(refreshed)
        return refreshed
    }

    private fun needsRefresh(token: SupabaseTokenSet): Boolean {
        if (token.refreshToken.isNullOrBlank()) return false
        if (token.issuedAtEpochSeconds == 0L) return true
        return clock() >= token.issuedAtEpochSeconds +
            (token.expiresIn - 30).coerceAtLeast(0)
    }

    private fun request(
        endpoint: String,
        method: String,
        accessToken: String? = null,
        body: String? = null,
    ): Response {
        val connection = openConnection(URL(endpoint)).apply {
            requestMethod = method
            setRequestProperty("apikey", publishableKey)
            accessToken?.let { setRequestProperty("Authorization", "Bearer $it") }
            connectTimeout = 10_000
            readTimeout = 10_000
            if (body != null) {
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
            }
        }

        try {
            if (body != null) {
                connection.outputStream.use { output ->
                    output.write(body.toByteArray(Charsets.UTF_8))
                }
            }
            val statusCode = connection.responseCode
            val stream = if (statusCode in 200..299) {
                connection.inputStream
            } else {
                connection.errorStream
            }
            val responseBody = stream?.bufferedReader()?.use { it.readText() }.orEmpty()
            return Response(statusCode, responseBody)
        } finally {
            connection.disconnect()
        }
    }

    private fun ensureSuccess(response: Response) {
        if (response.statusCode !in 200..299) {
            val detail = runCatching {
                val json = JSONObject(response.body)
                listOf("message", "msg", "error_description")
                    .asSequence()
                    .map { json.optString(it) }
                    .firstOrNull { it.isNotBlank() }
            }.getOrNull().orEmpty()
            throw IllegalStateException(
                if (detail.isBlank()) {
                    "Supabase Auth request failed with HTTP ${response.statusCode}."
                } else {
                    "Supabase Auth: $detail"
                },
            )
        }
    }

    private data class Response(
        val statusCode: Int,
        val body: String,
    )
}

File: apps/windows/Auth/SupabaseAuthClient.cs
using System.Net;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Andivum_Windows.Auth;

public sealed record AuthOperationResult(
    bool SessionCreated,
    bool EmailConfirmationRequired);

public sealed class SupabaseAuthClient
{
    private static readonly TimeSpan RefreshSkew = TimeSpan.FromSeconds(30);

    private readonly HttpClient httpClient;
    private readonly ITokenStore tokenStore;
    private readonly Uri authBaseUri;
    private readonly string publishableKey;
    private readonly SupabaseProfileClient profileClient;
    private readonly Func<DateTimeOffset> clock;

    public SupabaseAuthClient(
        HttpClient httpClient,
        ITokenStore tokenStore,
        Uri supabaseUrl,
        string publishableKey,
        Func<DateTimeOffset>? clock = null)
    {
        this.httpClient = httpClient;
        this.tokenStore = tokenStore;
        authBaseUri = new Uri($"{supabaseUrl.AbsoluteUri.TrimEnd('/')}/auth/v1/");
        this.publishableKey = publishableKey;
        this.clock = clock ?? (() => DateTimeOffset.UtcNow);
        profileClient = new SupabaseProfileClient(
            httpClient,
            tokenStore,
            supabaseUrl,
            publishableKey);
    }

    public TokenSet? CurrentSession => tokenStore.Read();

    public async Task<AuthOperationResult> SignInAsync(
        string email,
        string password,
        CancellationToken cancellationToken = default) =>
        await AuthenticateAsync(
            "token?grant_type=password",
            email,
            password,
            cancellationToken);

    public async Task<AuthOperationResult> SignUpAsync(
        string email,
        string password,
        CancellationToken cancellationToken = default) =>
        await AuthenticateAsync("signup", email, password, cancellationToken);

    public async Task<SessionResponse> GetCurrentSessionAsync(
        CancellationToken cancellationToken = default)
    {
        var token = tokenStore.Read() ??
            throw new InvalidOperationException("No saved session is available.");
        var hasRefreshed = false;

        if (NeedsRefresh(token))
        {
            await RefreshAsync(token, cancellationToken);
            hasRefreshed = true;
        }

        for (var attempt = 0; attempt < 2; attempt++)
        {
            try
            {
                return await profileClient.GetCurrentSessionAsync(cancellationToken);
            }
            catch (SessionUnauthorizedException) when
                (!hasRefreshed && !string.IsNullOrWhiteSpace(token.RefreshToken))
            {
                token = await RefreshAsync(token, cancellationToken);
                hasRefreshed = true;
            }
        }

        throw new InvalidOperationException(
            "Supabase session validation could not be completed.");
    }

    public async Task SignOutAsync(CancellationToken cancellationToken = default)
    {
        var token = tokenStore.Read();
        try
        {
            if (token is not null && !string.IsNullOrWhiteSpace(token.AccessToken))
            {
                using var request = new HttpRequestMessage(
                    HttpMethod.Post,
                    new Uri(authBaseUri, "logout"));
                request.Headers.Add("apikey", publishableKey);
                request.Headers.Authorization = new AuthenticationHeaderValue(
                    "Bearer",
                    token.AccessToken);
                using var response = await httpClient.SendAsync(request, cancellationToken);
            }
        }
        finally
        {
            tokenStore.Clear();
        }
    }

    private async Task<AuthOperationResult> AuthenticateAsync(
        string path,
        string email,
        string password,
        CancellationToken cancellationToken)
    {
        ValidateCredentials(email, password);
        using var request = new HttpRequestMessage(
            HttpMethod.Post,
            new Uri(authBaseUri, path))
        {
            Content = JsonContent.Create(new { email, password }),
        };
        request.Headers.Add("apikey", publishableKey);
        using var response = await httpClient.SendAsync(request, cancellationToken);
        var body = await response.Content.ReadAsStringAsync(cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            throw new InvalidOperationException(FormatAuthError(response.StatusCode, body));
        }

        var payload = JsonSerializer.Deserialize<SupabaseAuthPayload>(body) ??
            throw new InvalidOperationException("Supabase Auth returned no response.");
        if (string.IsNullOrWhiteSpace(payload.AccessToken))
        {
            return new AuthOperationResult(
                SessionCreated: false,
                EmailConfirmationRequired: path == "signup");
        }

        tokenStore.Save(payload.ToTokenSet(clock()));
        return new AuthOperationResult(
            SessionCreated: true,
            EmailConfirmationRequired: false);
    }

    private async Task<TokenSet> RefreshAsync(
        TokenSet current,
        CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(current.RefreshToken))
        {
            throw new InvalidOperationException("No Supabase refresh token is available.");
        }

        using var request = new HttpRequestMessage(
            HttpMethod.Post,
            new Uri(authBaseUri, "token?grant_type=refresh_token"))
        {
            Content = JsonContent.Create(new { refresh_token = current.RefreshToken }),
        };
        request.Headers.Add("apikey", publishableKey);
        using var response = await httpClient.SendAsync(request, cancellationToken);
        var body = await response.Content.ReadAsStringAsync(cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            throw new InvalidOperationException(FormatAuthError(response.StatusCode, body));
        }

        var payload = JsonSerializer.Deserialize<SupabaseAuthPayload>(body) ??
            throw new InvalidOperationException("Supabase refresh returned no response.");
        if (string.IsNullOrWhiteSpace(payload.AccessToken))
        {
            throw new InvalidOperationException("Supabase refresh returned no access token.");
        }

        var refreshed = payload.ToTokenSet(clock()) with
        {
            RefreshToken = string.IsNullOrWhiteSpace(payload.RefreshToken)
                ? current.RefreshToken
                : payload.RefreshToken,
        };
        tokenStore.Save(refreshed);
        return refreshed;
    }

    private bool NeedsRefresh(TokenSet token)
    {
        if (string.IsNullOrWhiteSpace(token.RefreshToken))
        {
            return false;
        }

        if (token.IssuedAt == default)
        {
            return true;
        }

        var refreshAt = token.IssuedAt + TimeSpan.FromSeconds(
            Math.Max(0, token.ExpiresIn - RefreshSkew.TotalSeconds));
        return clock() >= refreshAt;
    }

    private static void ValidateCredentials(string email, string password)
    {
        if (string.IsNullOrWhiteSpace(email))
        {
            throw new ArgumentException("Email is required.", nameof(email));
        }

        if (string.IsNullOrWhiteSpace(password))
        {
            throw new ArgumentException("Password is required.", nameof(password));
        }
    }

    private static string FormatAuthError(HttpStatusCode statusCode, string body)
    {
        var detail = string.Empty;
        try
        {
            var error = JsonSerializer.Deserialize<SupabaseErrorPayload>(body);
            detail = error?.Message ?? error?.Msg ?? error?.ErrorDescription ?? string.Empty;
        }
        catch (JsonException)
        {
            // Keep the user-facing error generic when the service returns malformed JSON.
        }

        return string.IsNullOrWhiteSpace(detail)
            ? $"Supabase Auth request failed with HTTP {(int)statusCode}."
            : $"Supabase Auth: {detail}";
    }

    private sealed record SupabaseAuthPayload(
        [property: JsonPropertyName("access_token")] string? AccessToken,
        [property: JsonPropertyName("refresh_token")] string? RefreshToken,
        [property: JsonPropertyName("token_type")] string? TokenType,
        [property: JsonPropertyName("expires_in")] int ExpiresIn,
        [property: JsonPropertyName("scope")] string? Scope)
    {
        public TokenSet ToTokenSet(DateTimeOffset issuedAt) => new(
            AccessToken ?? throw new InvalidOperationException("Supabase response has no access token."),
            RefreshToken,
            string.IsNullOrWhiteSpace(TokenType) ? "Bearer" : TokenType,
            ExpiresIn > 0 ? ExpiresIn : 3600,
            Scope,
            issuedAt);
    }

    private sealed record SupabaseErrorPayload(
        [property: JsonPropertyName("message")] string? Message,
        [property: JsonPropertyName("msg")] string? Msg,
        [property: JsonPropertyName("error_description")] string? ErrorDescription);
}
